Add new experiment processing/ reprocessing rules to gxa

pmb59 opened the pull request:
The aim of this PR is to add new rules and make of use existing ones (recalculations) to address processing of new experiments and reprocessing steps, in order to achieve a super-set of data analysis rules for GXA. This workflow now deals with previous atlas_prod scripts used before moving our systems to new EBI infrastructure.

The super-set of rules can now handle the following experiment types:
- [x] baseline
- [x] differential rnaseq
- [x] differential microarray

This PR includes: 
- logic that allows Snakemake rules collecting I/O files depending on the analysis goal (recalculations or reprocessing).
- (re)processing/recalculations only for selected specie/s
-  [updated differential expression strategy ](https://github.com/ebi-gene-expression-group/atlas-analysis/pull/1#event-6497594781)
- a number of Atlas scripts from production (which will be moved to [atlas-analysis](https://github.com/ebi-gene-expression-group/atlas-analysis) and then included as submodule here)
- communication to ISL db (only for new experiment or reprocessing)
- Major code changes/ new files are at:
```
- Snakefile
- Snakefile-reprocess
- Snakefile-sorting-hat
- bin/reprocessing_routines.sh
- bin/round_log2_fold_changes.R (reviewed)
- envs/isl-db.yaml
- run_sorting_hat_test_data.sh
```

Future merge requests will address:
- replacing generic bash routines with conda [atlas-bash-util](https://github.com/ebi-gene-expression-group/atlas-bash-util)  
- refactoring scripts using a submodule [atlas-analysis](https://github.com/ebi-gene-expression-group/atlas-analysis)
- adding proteomics and differential proteomics analysis rules
- add conditional structures that handle cases of optional/required parameters and files

pmb59 committed:
add comment

pmb59 committed:
Create var ACCESSIONS

If the env variable $ACCESSIONS is defined before running the script, only those accession will be considered for recalculations/reprocessing. If more than 1 accession is specified, then they should be separated by ":". e.g. `ACCESSIONS=E-MTAB-1234:E-MTAB-2544`

pmb59 committed:
add new reprocess rules

pmb59 committed:
add new reprocess rules

pmb59 committed:
add code to rule microarray_normalisation

pmb59 committed:
add rule check_configuration_xml and correct indenting

pmb59 committed:
Merge branch 'develop' into feature/reprocess-rules

pmb59 committed:
add comments

pmb59 committed:
This includes logic to check if the necessary XML files exist for an accession before commencing the reprocessing.

pmb59 committed:
append {acc}-raw-counts.tsv.undecorated to requested output

pmb59 committed:
Implementation of rule `copy_raw_gene_counts_from_isl`

pmb59 committed:
Add rule copy_normalised_counts_from_isl to reprocessing wf

pmb59 committed:
Add rule copy_transcript_files_from_isl

pmb59 committed:
Makes `rule copy_transcript_files_from_isl` to be run only when metric='tpms', as expected

pmb59 committed:
Add rule copy_transcript_relative_isoforms

pmb59 committed:
add localrules

pmb59 committed:
Add comment to rules copy_transcript_files_from_isl, copy_transcript_relative_isoforms.
Move rule rnaseq_qc

pmb59 committed:
Add rule rnaseq_qc

pmb59 committed:
Add rule quantile_normalise_expression.

pmb59 committed:
Conda env and scripts for rule quantile_normalise_expression

pmb59 committed:
Modify conda env (quantile.yaml) and other scripts necessary for rule quantile_normalise_expression.

pmb59 committed:
Include both metrics in rule quantile_normalise_expression

pmb59 committed:
Add rule summarize_expression

pmb59 committed:
Mark output file of rule quantile_normalise_expression as temp to maintain previous logic

pmb59 committed:
Add rule transcripts_na_check

pmb59 committed:
add rule quantile_normalise_transcripts

pmb59 committed:
Add rule summarize_transcripts

pmb59 committed:
Add rule generate_methods_baseline

pmb59 committed:
Modifies `rule generate_methods_baseline`. Corrects output name in Snakemake, further modifications to variables needed for gxa_generate_methods.pl

pmb59 committed:
Avoid workflow failure if isoform-file does not exist in rule copy_transcript_relative_isoforms. This is agreement with previous logic

pmb59 committed:
add rule decorate_expression_baseline and request output for Atlas Experiment Summary

pmb59 committed:
add container with ammonite

pmb59 committed:
add --use-singularity to snakemake call

pmb59 committed:
add metric to log in decorate rule

pmb59 committed:
update readme

pmb59 committed:
* polish rule decorate_expression_baseline
* split get_metrics() into two different functions for recalculations and re/processing. When processing of a new experiment, metrics cannot be based in the presence of files (fpkms or tmps) that the pipeline should create.

pmb59 committed:
request output for rules baseline_heatmap and link_baseline_heatmap

pmb59 committed:
request output for rule baseline_tracks

pmb59 committed:
add rule create_tracks_symlinks

pmb59 committed:
Fix rule summarize_transcripts. The rule is now dealing with situations in which quantile-normalised transcript file is not available

pcm32 commented on Snakefile:
```diff
@@ -16,9 +16,21 @@ def read_metadata_summary():
+#def get_isl_dir():
+#    return None
+def get_methods_template_baseline():
+    if 'methods_base' in config:
```
```suggestion
    if config['methods_base']:
```
I have learnt recently that this is apparently more "pythonic".

pcm32 replied:
Apologies, this is wrong, at it gives a missing key error if done!! Please ignore!

pcm32 commented on Snakefile:
```diff
@@ -94,6 +106,18 @@ def get_metrics():
         else:
             sys.exit("No metric available for baseline analyses.")
 
+def get_metrics_reprocess():
```
Could you elaborate on why you need this distinction? in principle both of them should use the same metrics. There is no reason why you would only recalculate or reprocess only certain metrics.

pmb59 replied:
Recalculations metrics is based on the availability of previously computed `{accession}-{metric}.tsv`, while for new experiment processing other logic should be implemented to get the metrics

pcm32 commented on Snakefile:
```diff
@@ -420,16 +445,64 @@ rule atlas_experiment_summary:
+
+# baseline_rnaseq_experiment
+
+rule check_configuration_xml:
```
I think that this rule can be obviated if another rules consumed the condensed SDRF explicitly.

pmb59 replied:
Done d16ae7f

pcm32 commented on Snakefile:
```diff
@@ -420,16 +445,64 @@ rule atlas_experiment_summary:
+        touch {output}
+        """
+
+rule check_factors_xml:
```
Idem here as in condensed

pmb59 replied:
Done d16ae7f

pcm32 commented on Snakefile:
```diff
@@ -420,16 +445,64 @@ rule atlas_experiment_summary:
+        set -e # snakemake on the cluster doesn't stop on error when --keep-going is set
+        exec &> "{log}"
+        source {workflow.basedir}/bin/reprocessing_routines.sh
+        expIslDir=$(find_exp_isl_dir {wildcards.accession})
```
I would prefer if we avoid this pattern of unearthing config mysteriously instead of passing the base ISL directory and working our way from there... if that works of course.

pmb59 replied:
Done, thanks!

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
-            cp $expIslDir/genes.{wildcards.metric}.htseq2.tsv {output.normalised_counts_undecorated}
-        elif [ -s "$expIslDir/genes.{wildcards.metric}.featurecounts.tsv" ]; then
-            cp $expIslDir/genes.{wildcards.metric}.featurecounts.tsv {output.normalised_counts_undecorated}
+            cp $expIslDir/genes.$metrictype.htseq2.tsv {output.normalised_counts_undecorated}
```
Some of these files might be large and maybe rsync is better here?

pmb59 replied:
done 6510ad0

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
         else
-        	echo "{params.exp_isl_dir}/transcripts.{wildcards.metric}.kallisto.tsv not found - skipping"
+            echo "$expIslDir/transcripts.tpm.kallisto.tsv not found - skipping"
+            exit 1
```
if we exit here this will be an error. I understand that some experiments (ie. controlled access or manually generated stuff) are allowed not to have transcripts. Or are you controlling which experiments will skip this rule elsewhere? If keeping the exit, I would remove the skipping part and just leave snakemake to fail on the missing file when copying.

pmb59 replied:
You are right, current logic should not break if file does not exist.

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
         if [ "$qcExitCode" -eq 2 ]; then
-        	echo "Experiment $expAcc has been disqualified due to insufficient quality, exiting"
+            echo "Experiment {wildcards.accession} has been disqualified due to insufficient quality, exiting"
             exit 0
```
I wonder here we should exit with an error or not.... otherwise we might think that the experiment got reprocessed, when in fact it didn't, and we will try to reload then.

pcm32 replied:
I would exit with an error here better.

pmb59 replied:
I agree, thanks

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+    conda: "envs/quantile.yaml"
+    log: "logs/{accession}-quantile_normalise_expression_{metric}.log"
+    input:
+        xml="{accession}-configuration.xml",
```
given that the configuration xml is being used here explicitly, I don't think that you need the rule upstream that asks for it to be present.

pcm32 replied:
I would also name this input `config_xml` or `config`, the format is rather the less relevant bit here.

pmb59 replied:
agree

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+            --configuration {input.xml}  \
+            < {input.qn_expression}  \
+            > {output.sum_expression}
+        if [ $? -ne 0 ]; then
+	        echo "ERROR: Failed to summarize gene expression in {wildcards.metric} for {wildcards.accession} " >&2
+	        exit 1
+        fi
```
```suggestion
```
I would remove this bit.

pcm32 replied:
you probably won't get there anyway I think.

pmb59 replied:
thanks @pcm32 why do you think `perl gxa_summarize_expression.pl` won't produce errors?

pcm32 replied:
Won't `set -e` mean that it will stop and exit at the first failure? If that is the case, then you will never reach that if.

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+            echo "transcripts NA check -  executed for {input.transcripts} "
+        else
+            echo "$expIslDir/transcripts.raw.kallisto.tsv not found for {wildcards.accession} - skipping rule_transcripts_na_check for {input.transcripts}"
+            #exit 1
```
```suggestion
```

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+        if [ -s {params.transcripts} ] ; then
+
+            {workflow.basedir}/bin/quantile_normalize.sh -c {input.xml} -s {params.transcripts} -d {params.qntranscripts} -b {workflow.basedir}/bin
+
+            if [ $? -ne 0 ]; then
+                echo "ERROR: Failed to quantile normalize for {params.transcripts}  " >&2
+                exit 1
+            fi
```
```suggestion
```

pmb59 replied:
idem why do you think ` quantile_normalize.sh` won't produce errors?

pcm32 replied:
`set -e`

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+                --configuration {input.xml}  \
+                < {params.qn_transcripts}  \
+                > {params.agg_transcripts}
+            if [ $? -ne 0 ]; then
+                echo "ERROR: Failed to summarize transcript expressions TPMs for {params.qn_transcripts}  " >&2
+                exit 1
+            fi
```
```suggestion
```

pmb59 replied:
thanks @pcm32 This check was inherited from previous code. Why do you think it's unnecessary?

pcm32 replied:
`set -e` won't let you reach this.

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+        """
+
+
+rule generate_methods_baseline:
```
Maybe we should be explicit in all rule names if they are rna_seq specific, as we will soon need to add the proteomics ones. In this case, this one is clearly different proteomics to rna_seq (sources of these methods files will differ).

pmb59 replied:
agree, renamed to `rule generate_methods_baseline_rnaseq`

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+
+        perl {workflow.basedir}/bin/gxa_generate_methods.pl "$expIslDir/irap.versions.tsv" {wildcards.accession} {params.organism} {params.template} "${{baseline_mapper:?}}" "${{baseline_quantMethod:?}}" "${{de_mapper:?}}" "${{de_
+quantMethod:?}}" "${{de_deMethod:?}}" > {output.methods}
+
+        if [ $? -ne 0 ]; then
+            echo "ERROR: Failed to generate analysis methods for {wildcards.accession}" >&2
+            exit 1
+        fi
```
```suggestion
```

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+        decoratedFileLength=$(wc -l "$decoratedFile.swp" | cut -f 1 -d ' ' )
+        if [ -s "$decoratedFile.swp" ] && [ "$decoratedFileLength" -gt 1 ]; then
+            mv $decoratedFile.swp $decoratedFile
+        else
```
I suspect that this won't be reached, I would remove it.

pcm32 commented on Snakefile:
```diff
@@ -440,72 +513,452 @@ rule copy_raw_gene_counts_from_isl:
+        fi
+
+        ln -s {input.bedGraph} {params.output}
+
+        if [ $? -ne 0 ]; then
+          echo "ERROR: creating symkink for {wildcards.accession} {wildcards.metric} {wildcards.assay_id}"
+          exit 1
+        fi
+
```
```suggestion
```

pmb59 replied:
happy to remove this one!

pcm32 reviewed (commented):
Some initial thoughts and suggestions to improve this. Good work on getting this to work  @pmb59 !

pmb59 committed:
remove rules check_configuration_xml and check_factors_xml

pmb59 committed:
Switch `cp` to `rsync -avz`

pmb59 committed:
remove comment exit 1

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
address pcm32 comments

pmb59 committed:
Merge branch 'feature/reprocess-rules' of https://github.com/ebi-gene-expression-group/bulk-recalculations into feature/reprocess-rules

pmb59 committed:
add definition of rules for differential analysis

pmb59 committed:
add rule decorate_transcripts_baseline

pmb59 committed:
Replace find_exp_isl_dir() with path retrieved directly from config

pmb59 committed:
Fix bug (`--aggregate quartiles` should not occur in summarize transcripts) and possible source of conflict (`transcripts_expr_values_check.R` input and output is same file. Now quantile normalisation will happen only after rule transcripts_na_check has finished)

pmb59 committed:
add rule differential_statistics_rnaseq

pmb59 committed:
add `TMPDIR` to rule differential_statistics_rnaseq via params

pmb59 committed:
add rule check_mvaPlot_rnaseq

pmb59 committed:
add rule round_log2_fold_changes_rnaseq

pmb59 committed:
add rule generate_methods_differential_rnaseq

pmb59 committed:
add {acc}-atlasExperimentSummary.Rdata to list of outputs for differential rna-seq.
add rule decorate_differential_rnaseq.

pmb59 committed:
add fixes to rule decorate_differential_rnaseq

pmb59 committed:
request output in `rule percentile_ranks` for differential rna-seq

pmb59 committed:
add differential gsea output

pmb59 committed:
Avoid incomplete matrices in {acc}-percentile-ranks.tsv files after the `data.table` developers (R package that generates out percentile ranks) changed the default to `na=""` in `fwrite`

pmb59 committed:
add output for rule differential_tracks (rna-seq)

pmb59 committed:
- Improve logic: add functions as input to certain rules to differentiate between Reprocessing and Recalculations
- get analysis goal (recalculations or reprocess) in config from the sorting-hat
- define global variable experiment type

pmb59 committed:
add output to Snakefile-reprocess

pmb59 committed:
add rule get_normalized_expressions (differential microarrays)

pmb59 committed:
add fixes to rule get_normalized_expressions_microarray

pmb59 committed:
Add rule check_normalized_expressions_microarray

pmb59 committed:
add `rule microarray_qc`

pmb59 committed:
add check_normalized_expressions_microarray to local rules

pmb59 committed:
add logic to check that one QC subfolder is created for each array_design of the accession

pmb59 committed:
rename rules for differential microarray analysis

pmb59 committed:
add rule generate_methods_differential_microarray

pmb59 committed:
add rule decorate_norm_expr_microarray

pmb59 committed:
add rule merge_probe_ids_microarray

pmb59 committed:
save a copy of initial normalized expressions in differential microarrays (unmerged probe ids)

pmb59 committed:
add rule differential_statistics_microarray

pmb59 committed:
add rule check_nas_microarray

pmb59 committed:
add rule round_log2_fold_changes_microarray

pmb59 committed:
Replace perl script for rounding logic

pcm32 commented on bin/round_log2_fold_changes.R:
```diff
@@ -0,0 +1,48 @@
+args <- commandArgs( TRUE )
+
+if( length( args ) != 3 ) {
+  stop( "\nUsage:\n\tround_log2_fold_changes.R <experiment_type> <unrounded_analytics_filename> <rounded_analytics_filename> \n\n" )
```
When using too many positional arguments please use R optparse instead with flags.

pmb59 replied:
done, thanks

pcm32 commented on bin/round_log2_fold_changes.R:
```diff
@@ -0,0 +1,48 @@
+	setcolorder(unrounded_dt, reordering )
+} 
+# Write out
+write.table( unrounded_dt, file = rounded, row.names=FALSE, quote=FALSE, sep="\t" )
```
Are we 100% sure that the output here is the same in every formatting respect as generated by the previous perl script? Might be good to run this against a couple of different experiments for each technology and type to check. Otherwise something else down the line will break.

pmb59 replied:
I noticed the perl script was forcing the `NAs` to be `0` after rounding. I will replicate that behaviour for consistency. Otherwise, same formatting for differential rna-seq and microarrays (not so well tested for proteomics,  but I will)

pmb59 replied:
PS. Proteomics will be addressed in a different PR

pcm32 reviewed (commented):
Looks good in general, besides optparse, please just make really sure that the output is entirely equivalent in terms of formatting to the previous one (this review is only about this round_log2_fold script).

pmb59 committed:
replace `NAs` with `0s` in log-fold-changes before rounding.

pmb59 committed:
use the R package optparse in round_log2_fold_changes.R and make the conversion of NAs to 0s faster

pmb59 committed:
add Snakefile-proteomics

pmb59 committed:
Update Snakefile-recalculations

pmb59 committed:
Update Snakefile-reprocess

pmb59 committed:
* add rule decorate_differential_microarray.
* split decorate routines for microarrays (decorate_microarray_file / decorate_normalized_file).

pmb59 committed:
request percentile ranks output for differential microarrays, and necessary logic.

pmb59 committed:
request outputs for differential GSEA and diff tracks in microarray analysis.

pmb59 committed:
add local rule to delete intermediate files generated during microarray reprocessing.

pmb59 committed:
Snakemake input functions need to return a string or a list of strings.

pcm32 commented on Snakefile-reprocess:
```diff
@@ -1,26 +1,170 @@
+    """
+    Parse xml config file here.
+    """
+    from xml.dom import minidom
```
In general I would avoid DOM based approaches for parsing XML as they load the entire object in memory, but if these XML files are not that big for microarrays, this should be fine.

pmb59 committed:
Fix to the error occurring when reading array_designs from the config xml files.

When the xml is in multiline format
```
        <array_design>
        A-AFFY-130
        </array_design>
```
the output is `['\n             A-AFFY-130\n        ']` and the spaces and newlines should be removed from the str.

pmb59 committed:
Define two final rules for all newly reprocessed exps.

pmb59 committed:
add finalisaling rules to cp data to atlas_exps and get the condensed SDRF

pmb59 committed:
correct syntax error

pmb59 committed:
add `atlas_env()` to reprocessing routines

pinin4fjords commented on bin/reprocessing_routines.sh:
```diff
@@ -207,3 +207,12 @@ applyFixes() {
+
+# Returns prod or test, depending on the Atlas environment in which the script calling it is running
+# It is assuming that all atlasinstall_<env>s are under ${ATLAS_PROD}/sw (it will fail otherwise)
+atlas_env() {
```
Sorry, should have made the suggestion earlier. But if you're moving away from atlas-prod maybe functions like this could be called from atlas-bash-util? https://github.com/ebi-gene-expression-group/atlas-bash-util/blob/2d9c0893de2e1fe98c67f320f74466c2d2351646/generic_routines.sh#L71

With atlas-bash-util installed this call would become:

```
atlas-bash-util atlas_env
```

pmb59 replied:
thanks @pinin4fjords, this will be addressed globally in a different PR as indicated in the description.

pmb59 committed:
update differential expression strategy

pmb59 committed:
add reshape2 dependency to conda env for diff statistics

pinin4fjords commented on envs/differential-stats.yaml:
```diff
@@ -6,6 +6,7 @@ dependencies:
  - perl-atlas-modules >=0.3.1
  - r-atlas-internal>=1.8
  - r-ggplot2
+ - r-reshape2
```
Just a note for as and when this starts being used with containers- there is already a container for this built from https://github.com/ebi-gene-expression-group/atlas-containers/tree/master/atlas-analysis-base

pmb59 replied:
Thanks @pinin4fjords  I have added a note. Not sure at the moment how the version restriction will `affect perl-atlas-modules` and `r-atlas-internal`, but it would be good to remove those restrictions.

pcm32 commented on Snakefile-sorting-hat:
```diff
@@ -88,7 +88,41 @@ def get_check_sp_file():
+    if 'isl_dir' in config:
+        return config['isl_dir']
+    else:
+        return None
```
Will the calling methods of these getters survive if the config is not set? If there is no other case, you might as well leave the direct config call and let snakemake fail directly. I would use these methods when there is a reasonable default for when config is not set, or if the rule actually continues to make sense when they are not defined.

pmb59 replied:
thanks, done here [bf080aa](https://github.com/ebi-gene-expression-group/bulk-recalculations/pull/24/commits/bf080aa187bd5ca73b69de483c49674d2439258e)

pcm32 reviewed (commented):
LGTM in general. I did skip a lot of scripts because I suspect that most of them are simply copies from atlas-prod and that is not what we need to be reviewing here. But it would be good if you add a comment (in a review manner) on every script that is completely new (not copied from Atlas), to give those a more detailed inspection. Thanks!

pmb59 committed:
Allows to run reprocessing/recalculations on a list user-defined SPECIES (organisms). The species provided should be separated by ':', e.g. `SPECIES=homo_sapiens:rattus_norvegicus`.
If a list of studies is defined (in ACCESSIONS) then acc takes priority and SPECIES is left empty.

pmb59 committed:
Define rule delete_experiment

pmb59 committed:
add submodule with isl

pmb59 committed:
add rule add_runs_to_db (isl) for rna-seq experiments (baseline and differential)

pmb59 committed:
add get_db_params()

pmb59 committed:
adds Oracle db params.
Some fixes on the logic to distinguish (re)processing/recalculations operations.

pmb59 committed:
This commit impedes `rule atlas_experiment_summary` starting its execution during reprocessing before necessary files are present.

pmb59 committed:
Include rule and associated logic to avoid execution of upstream rules when `forceall=true` in baseline rna-seq recalculations.

pmb59 committed:
update run_sorting_hat script

pmb59 committed:
add `rule copy_experiment_from_analysis_to_atlas_exps` and related routines

pmb59 committed:
add rule `get_magetab_for_experiment` to generate condensed SDRFs with Zooma mappings.

pmb59 committed:
correct indent, edit description of delete rule

pmb59 commented:
> LGTM in general. I did skip a lot of scripts because I suspect that most of them are simply copies from atlas-prod and that is not what we need to be reviewing here. But it would be good if you add a comment (in a review manner) on every script that is completely new (not copied from Atlas), to give those a more detailed inspection. Thanks!

thanks @pcm32 in general most of atlas-prod scripts required small fixes during testing, which did not affect the overall functionality, but not simple copies. I will highlight in the description what the most important scripts and file changes are.

pmb59 committed:
add zooma_exclusion YAML file to Snakemake config and associate wf control logic.

pmb59 committed:
leave the direct config call and let Snakemake fail if a required fail is missing

pmb59 committed:
Few edits to enable `rule rnaseq_qc` (qc steps for RNA-seq experiments)

pmb59 commented on Snakefile:
```diff
@@ -420,92 +602,1307 @@ rule atlas_experiment_summary:
+        """
+
+######################################################
+# Potential new rule, to be implemented
+
+rule delete_experiment:
+    """
+    Delete experiment from Atlas.
+    It should also remove its RUN IDs from ISL db.
+    add goal='delete' in addition to 'recalculations' and 'reprocess'
+    """
```
@pcm32 I've noticed this is something done by the curation team under current _Updating existing experiments_ SOP. Could you confirm we should not implement this here?

pcm32 replied:
Currently this happens via a Jenkins job, which probably today is using a call to the web application besides some deletions... I'm not sure how we are doing and to what extent is done (solr, db, file system), but we will need a system for this. I suspect that currently the easiest is to continue to rely on the web application call until we devise a web-cli call for this... or have we already?

pcm32 replied:
ahhh, ok, no implementation here yet. I would leave it (as in not doing it - remove it if you want) in the meantime. The Jenkins job is surely relying on the web application for this.

pcm32 commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "/Users/pmadrigal/bulk-recalculations/isl"]
```
This submodule should be added in a generic manner (relative paths and not bound to a user).

pcm32 replied:
In this format, it won't work on other user's machines I expect.

pmb59 replied:
This was changed here [7806fc0](https://github.com/ebi-gene-expression-group/bulk-recalculations/commit/7806fc051848f78b93be23e5808502479dc3b8ca)

pmb59 committed:
rm rule delete_experiment

pmb59 committed:
get DESeq2 package version from conda env used for differential expression analysis.

pmb59 committed:
Add ISL_GENOMES to the wf config, and as input param in bin/gxa_generate_methods.pl (to replace get_genome_references_config )

pmb59 committed:
Fix .gitmodules

pmb59 committed:
Add logic to transfer the IRAP versions .mk file (for bin/gxa_generate_methods.pl) from the singularity container if it is not present.

pmb59 committed:
Implement get_metrics_reprocess() based on content of ISL folders.

pcm32 commented on Snakefile:
```diff
@@ -94,6 +142,51 @@ def get_metrics():
+            metrics_reprocess.append('fpkms')
+
+        if not metrics_reprocess:
+            sys.exit("No metrics for reprocessing found in isl path.")
```
Is this exiting with a non zero status to show failure? can we set it like that otherwise? Thanks.

pmb59 replied:
Yes. If a non-integer is passed, `None` is equivalent to passing 0, and any other object (such as the string here) is printed to stderr and results in an exit code of 1.

pcm32 commented on Snakefile:
```diff
@@ -94,6 +142,51 @@ def get_metrics():
+        if i in config:
+            db_params.append( config[i] )
+        else:
+            print(f" Missing db param: {i}")
```
```suggestion
            print(f" Missing ISL db param: {i}")
```

pcm32 commented on Snakefile:
```diff
@@ -94,6 +142,51 @@ def get_metrics():
+            db_params.append( config[i] )
+        else:
+            print(f" Missing db param: {i}")
+            #sys.exit()
```
```suggestion
            sys.exit(1)
```

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+    """
+    if config['goal'] == 'reprocess':
+        if experiment_type=='rnaseq_mrna_differential':
+            return [ 'logs/'+wildcards['accession']+'-decorate_differential_rnaseq.done' ]
```
```suggestion
            return [ f"logs/{wildcards['accession']}-decorate_differential_rnaseq.done" ]
```
prefer f strings in general

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+            inputs = []
+            arr_designs=get_array_design_from_xml()
+            for s in arr_designs:
+                inputs.append( 'logs/'+wildcards['accession']+'_'+s+'-decorate_differential_microarray.done' )
```
f strings here please as well.

pmb59 replied:
done 1fdfe63

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+            return None
+    if config['goal'] == 'recalculations':
+        # No input file needed - trick to force rule execution
+        return wildcards['accession']+'.metadata_summary.yaml'
```
```suggestion
        return f"{wildcards['accession']}.metadata_summary.yaml"
```

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+    """
+    if config['goal'] == 'reprocess':
+        if experiment_type=='rnaseq_mrna_differential':
+            return [ 'logs/'+wildcards['accession']+'-decorate_differential_rnaseq.done' ]
```
```suggestion
            return [ f"logs/{wildcards['accession']}-decorate_differential_rnaseq.done" ]
```

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+            inputs = []
+            arr_designs=get_array_design_from_xml()
+            for s in arr_designs:
+                inputs.append( 'logs/'+wildcards['accession']+'_'+s+'-decorate_differential_microarray.done' )
```
```suggestion
                inputs.append( f"logs/{wildcards['accession']}_{s}-decorate_differential_microarray.done" )
```

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+            inputs = []
+            arr_designs=get_array_design_from_xml()
+            for s in arr_designs:
+                inputs.append( wildcards['accession']+'_'+s+'-normalized-expressions.tsv.undecorated' )
```
f string

pmb59 replied:
8ff665f

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+            arr_designs=get_array_design_from_xml()
+            for s in arr_designs:
+                inputs.append( wildcards['accession']+'_'+s+'-log-fold-changes.tsv.undecorated' )
+                inputs.append( wildcards['accession']+'_'+s+'-average-intensities.tsv.undecorated' )
```
f string

pmb59 replied:
8ff665f

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
+    if config['goal'] == 'reprocess':
+        inputs = get_outputs()
+        inputs.remove('logs/'+wildcards['accession']+'-copy_experiment_from_analysis_to_atlas_exps.done')
+        inputs.remove('logs/'+wildcards['accession']+'-get_magetab_for_experiment.done')
```
f string

pmb59 replied:
82ea574

pcm32 commented on Snakefile:
```diff
@@ -164,18 +256,113 @@ def get_mem_mb(wildcards, attempt):
 
 
-localrules: check_differential_gsea, link_baseline_coexpression, link_baseline_heatmap
+localrules: check_differential_gsea, link_baseline_coexpression, link_baseline_heatmap, copy_raw_gene_counts_from_isl, copy_normalised_counts_from_isl, copy_transcript_files_from_isl, copy_transcript_relative_isoforms, create_tracks_symlinks, check_mvaPlot_rnaseq, check_normalized_expressions_microarray, delete_intermediate_files_microarray, touch_inputs_baseline
```
I wonder if we should leave the large copies to be sent over to the cluster? Otherwise you will have a bottle neck here which will delay in sending other jobs possibly. If copies are small, I would keep them here, but for large copies I would send them as normal jobs (not localrules).

pmb59 replied:
thanks @pcm32  that makes sense 2931bf1

pcm32 commented on Snakefile:
```diff
@@ -377,16 +588,17 @@ rule baseline_heatmap:
         exec &> "{log}"
         {workflow.basedir}/bin/generateBaselineHeatmap.R --configuration {wildcards.accession}-configuration.xml \
-		--input  {input.expression} \
+		--input  {params.expression} \
```
This looks a bit weird, we have expression in both params and input. If the expression is a file (or set of files), we should use input, as being in params snakemake doesn't make the typical checks done for a file (input is always a file unless tagged as a directory I think).

pmb59 replied:
thanks @pcm32 I've removed the params 8ee9634

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+# baseline_rnaseq_experiment
+
+rule add_runs_to_db:
+    """
+    Get run ids from config file and add them to isl db.
+    """
+    conda: "envs/isl-db.yaml"
+    log: "logs/{accession}-add_runs_to_db.log"
+    input:
+        config_xml="{accession}-configuration.xml"
     params:
-        exp_isl_dir=get_isl_dir()
+        db_params=get_db_params()
+    output:
+        temp("logs/{accession}-add_runs_to_db.done")
+    shell:
+        """
+        set -e # snakemake on the cluster doesn't stop on error when --keep-going is set
+        exec &> "{log}"
+        export TNS_ADMIN={params.db_params[0]}/network/admin
+        export LD_LIBRARY_PATH={params.db_params[0]}/lib:$LD_LIBRARY_PATH
+        export PATH={params.db_params[0]}/bin:$PATH
+        export PYTHON_USER={params.db_params[1]}
+        export PYTHON_CONNECT_STRING={params.db_params[2]}
+        export PYTHON_PASSWORD={params.db_params[3]}
+
+        python {workflow.basedir}/isl/db/scripts/get_run_ids_atlas_prod.py {input.config_xml}
+        if [ $? -ne 0 ]; then
+	        echo "ERROR: Failed to parse atlas config file and get run ids for {wildcards.accession} " >&2
+	        exit 1
+        fi
+        touch {output}
+        """
```
I wonder if this rule should be part of reprocess/recalculations? This seems like an undesirable circular loop between our processes. What do we need this for currently?

pmb59 replied:
As discussed today, it depends on whether the cron job deals with this task or not. In the worst case scenario, keeping the rule is harmless:
- it will insert the run_ids collected into LIBRARIES_ATLAS, if they have not been inserted before.
- otherwise it does nothing, except some qc checks. For instance if there are no valid run_ids in the experiment will raise an ERROR. In that case it is also convenient to stop running the workflow. I will also trigger WARN if there are duplicated Run ids or with unexpected accessions.

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        [ ! -z $expIslDir+x ] || (echo "snakemake param exp_isl_dir needs to defined in rule" && exit 1)
+
+        if [ -s "$expIslDir/transcripts.tpm.kallisto.tsv" ] ; then
+            rsync -avz $expIslDir/transcripts.tpm.kallisto.tsv {output.transcripts}
         else
-        	echo "{params.exp_isl_dir}/transcripts.{wildcards.metric}.kallisto.tsv not found - skipping"
+            echo "$expIslDir/transcripts.tpm.kallisto.tsv not found - skipping"
         fi
```
Why is this explicit on tpm and not parametrised by `{metric}`?

pmb59 replied:
Thanks @pcm32, this is an important point. This code was inherited from [here](https://github.com/ebi-gene-expression-group/atlas-prod/blob/develop/db/scripts/import_baseline_rnaseq_experiment.sh#L39-L51). Should we bring `transcripts.fpkm.kallisto.tsv` to data production? 
Pinning @pinin4fjords in case he wants to comment.

pcm32 replied:
I guess that if we have fpkms for transcripts we should bring it over, right? Can you check if we have them in general (I guess in noah) among the ISL results?

pcm32 replied:
In any case, if this wasn't part of the original code, I'm happy for us to leave this like this towards closing this monster PR and open an issue and then resolve it in a smaller PR.

pinin4fjords replied:
I'm afraid that I don't know what the rationale was of our predecessors with the original code, so I don't have a useful comment to make here.

pcm32 replied:
Issue: https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/32

pmb59 replied:
thanks - `transcripts.fpkm.kallisto.tsv` are available, but have not been reported by Atlas so far (see the FTP). I will add a new feature in a different branch to allow this.

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        expIslDir={params.isl_dir}/{wildcards.accession}/{params.organism}
+        echo "ISL dir: $expIslDir"
+
+        [ ! -z $expIslDir+x ] || (echo "snakemake param exp_isl_dir needs to defined in rule" && exit 1)
+
+        if [ -s "$expIslDir/transcripts.tpm.kallisto.tsv" ] ; then
+            rsync -avz $expIslDir/transcripts.tpm.kallisto.tsv {output.transcripts}
         else
-        	echo "{params.exp_isl_dir}/transcripts.{wildcards.metric}.kallisto.tsv not found - skipping"
+            echo "$expIslDir/transcripts.tpm.kallisto.tsv not found - skipping"
         fi
         """
 
 rule copy_transcript_relative_isoforms:
-    params:
-        exp_isl_dir=get_isl_dir()
+    """
+    Copy transcripts relative isoform usage files.
+    If file does not exist for an accession, this rule can be skipped.
+    """
+    log: "logs/{accession}-copy_transcript_relative_isoforms.log"
     output:
-        transcripts_relative_isoforms="{accession}-transcripts.riu.tsv"
+        temp("logs/{accession}-copy_transcript_relative_isoforms.done")
+    params:
+        transcripts_relative_isoforms="{accession}-transcripts.riu.tsv",
+        organism=get_organism(),
+        isl_dir=get_isl_dir()  
     shell:
         """
+        set -e # snakemake on the cluster doesn't stop on error when --keep-going is set
+        exec &> "{log}"
+        expIslDir={params.isl_dir}/{wildcards.accession}/{params.organism}
+        echo "ISL dir: $expIslDir"
+
+        [ ! -z $expIslDir+x ] || (echo "snakemake param exp_isl_dir needs to defined in rule" && exit 1)
+
         if [ -s "$expIslDir/transcripts.riu.kallisto.tsv" ] ; then
-            cp $expIslDir/transcripts.riu.kallisto.tsv {output.transcripts_relative_isoforms}
+            rsync -avz $expIslDir/transcripts.riu.kallisto.tsv {params.transcripts_relative_isoforms}
```
This pattern seems a bit hacky, couldn't we rely on the presence / absence of the input `$expIslDir/transcripts.riu.kallisto.tsv` to decide the output?

pmb59 replied:
See issue https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/33

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+    log: "logs/{accession}-rule-transcripts_na_check_{metric}.log"
+    params:
+        organism=get_organism(),
+        isl_dir=get_isl_dir()  
```
Shouldn't this be a directory input? I think you can do something like:
```
   input:
       isl_dir=directory(get_isl_dir())
```
and then snakemake will make sure that this exists as an input, I think. Unless that this is undesirable.

pmb59 replied:
TTBOMK, this keyword is only valid for outputs in Snakemake.

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        touch {output}
+        """
+
+rule summarize_transcripts:
+    """
+    Summarize transcript expression in tpms, if the file exists,
+    either into median per biological replicate or into quartile per assay group.
+    """
+    conda: "envs/perl-atlas-modules.yaml"
+    log: "logs/{accession}-summarize_transcripts_{metric}.log"
+    input:
+        xml="{accession}-configuration.xml",
+        getqn=rules.quantile_normalise_transcripts.output
+    output:
+        temp("logs/{accession}-summarize_transcripts_{metric}.done")
+    params:
+        qn_transcripts="{accession}-transcripts-{metric}.tsv.undecorated.quantile_normalized",
+        agg_transcripts="{accession}-transcripts-{metric}.tsv.undecorated.aggregated"
+    shell:
```
I suspect that it is a bit hacky again to use files as parameters like this, but I see why you do it. What is the "Snakemaky" way of doing this @pinin4fjords ?

pmb59 replied:
See https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/33

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            fi
+        else
+            echo "The file {params.irap_versions} already exists."
+        fi
```
Maybe better to have a rule for this and use the resulting {irap_versions} explicitly as an input (from a central location)? It will only run once as well.

pmb59 replied:
Done 61a9fd1. This logic was necessary to avoid:

> Not all output, log and benchmark files of rule get_irap_versions_file contain the same wildcards. This is crucial though, in order to avoid that two or more jobs write to the same file.

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        echo $geneNameFile
+
+        test -s "$geneNameFile" || (  >&2 echo "$0 gene name file not found: $geneNameFile" ; exit 1 )
+        #test -s "$FPKMexpressionsFile" -o -s "$TPMexpressionsFile" || (  >&2 echo "$0 no data files for $e" ; exit 1 )
```
```suggestion
```
if we are sure that we don't want this anymore.

pmb59 replied:
done 8e122dd

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        --geneNameFile "$geneNameFile" \
+        --source {input.expression} \
+        | awk 'NR == 1; NR > 1 {{print $0 | "sort -n"}}' \
+        > $decoratedFile.swp
```
```suggestion
            --geneNameFile "$geneNameFile" \
            --source {input.expression} \
            | awk 'NR == 1; NR > 1 {{print $0 | "sort -n"}}' \
            > $decoratedFile.swp
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            decoratedFile={params.decotranscripts} 
+
+            # Ammonite REPL & Script-Runner
+            #export JAVA_OPTS="-Xmx3000M"
```
I remember that some files required more memory for decoration, was this commented in the code?

pmb59 replied:
thanks @pcm32  The flag `   export JAVA_OPTS="-Xmx3000M"` was used in the old setup. Actually only for baseline transcript decoration [here](https://github.com/ebi-gene-expression-group/atlas-prod/blob/develop/bioentity_annotations/decorate_routines.sh#L57). Ammonite man says the following: 

> Set a default heap limit of 500 megabytes to the amm standalone executable, so it no longer grows to the "default" of 1/4 your system memory (typically multiple gigabytes) unnecessarily. If you need to give it more memory, start it with JAVA_OPTS=-Xmx1500M amm

I don't think we have experienced problems in this rule, but I will provide ammonite available memory on the workflows (which depends on Snakemake attempts), just in case

pmb59 replied:
done 971c0d4

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            if [ -s "$decoratedFile.swp" ] && [ "$decoratedFileLength" -gt 1 ]; then
+                mv $decoratedFile.swp $decoratedFile
+            else
+                echo "ERROR: decorate_transcript_baseline_file for {wildcards.accession} and {wildcards.metric}"
+	            exit 1
```
```suggestion
                echo "ERROR: decorate_transcript_baseline_file for {wildcards.accession} and {wildcards.metric}"
	        exit 1
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        fi
+
+        ln -s {input.bedGraph} {params.output}
+        touch {output}
```
why not?

```suggestion
    output: "{accession}.{assay_id}.genes.expressions.bedGraph"
    shell:
        """
        set -e # snakemake on the cluster doesn't stop on error when --keep-going is set
        exec &> "{log}"

        # remove if any old bedgraph files
        rm -f {output} 

        ln -s {input.bedGraph} {output}
```

and make sure that this output is added to the list of required outputs for the all rule, and remove the .done ?

pmb59 replied:
@pcm32 I did that to avoid to avoid the uncertainty error in Snakemake produced because the output file has no wildcard `{metric}` on it. Hence the required output was: 
```
        outputs.extend(expand(f"{config['accession']}"+".{a_id}.create_tracks_symlinks_{metric}.done",
                            a_id=get_assay_ids(),
                            metric=["tpms"]))
```
Now, I have removed this wildcard, as only TPMs are required, 68be696

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+	        echo "ERROR: Failed to generate differential expression statistics and MvA plots for {wildcards.accession}" >&2
+	        rm -rf *.png {wildcards.accession}-analytics.tsv.undecorated
+	        exit 1
+        fi
```
```suggestion
```
this will never be reached due to set -e

pmb59 replied:
d95eb9c

pcm32 commented on Snakefile:
```diff
@@ -16,9 +16,54 @@ def read_metadata_summary():
+    else:
+        return None
+
+def get_tmp_dir():
+    if 'tmp_dir' in config:
+        return config['tmp_dir']
+    else:
+        return None
```
I see this function used a lot, is there a need for all the tools to use the same tmp dir (so that is why it is being set from upstream) or in most case you just one independent tmp dirs (what I would guess). If that is the case, I would avoid taking this all the way to snakemake and simply use `TMPDIR=$(mktemp --directory)` on the shells (that will create a local temp directory on the machine where this is running). You need to remember to delete it of course.

pmb59 replied:
Unfortunately, currently mktemp won't create a temp dir name in appropriate codon path. But I have included this option in case we do not want a custom TMP_DIR 18bd548

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        echo $TMPDIR"/tmp"
+        if [ ! -d "$TMPDIR"/tmp ]; then
+            mkdir $TMPDIR/tmp
+        fi
```
There is a lot of logic here for a temp directory, I would get rid of the params.tmp_dir and simply use:
```suggestion
        TMPDIR=$(mktemp --directory)
        echo $TMPDIR
```
is this used internally in the scripts below? I don't see it used at least at the bash level.

pcm32 replied:
the same in other occurrences of the same, if it works.

pmb59 replied:
Currently we depend on a custom temp_dir location in codon (see comment above), but I have moved basic logic to the common routines script f678934

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        set -e # snakemake on the cluster doesn't stop on error when --keep-going is set
+        exec &> "{log}"
+
+        if [ -s "{wildcards.accession}-{wildcards.contrast_id}-mvaPlot.png" ] ; then
+            echo "mvaPlot for {wildcards.accession} and {wildcards.contrast_id} successfully generated during differential rna-seq statistics: {wildcards.accession}-{wildcards.contrast_id}-mvaPlot.png"
+        else
+            echo "ERROR: mvaPlot for {wildcards.accession} and {wildcards.contrast_id} not produced during differential rna-seq statistics."
+            exit 1
```
Why not declare this file for accession/contrast_id as output of the rule that is generating them?

pmb59 replied:
See issue https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/33

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+	        echo "ERROR: Failed to round to one decimal place log2fold changes in {input} " >&2
+	        rm -rf {params.intermediate_rounded}
+	        exit 1
+        fi
```
```suggestion
```

will not be reached.

pmb59 replied:
5198675

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+	        rm -rf {params.intermediate_rounded}
+	        exit 1
+        fi
+        mv {input} {output.unrounded}
+        mv {params.intermediate_rounded} {input}
```
This means that this step will always be redone, is this intended? This logic is a bit strange.

pmb59 replied:
Fully agree, but I respected the Atlas logic defined previously [here](https://github.com/ebi-gene-expression-group/atlas-prod/blob/develop/db/scripts/import_differential_rnaseq_experiment.sh#L59-L60). 
In my experience this rule is not redone during reprocessing, at least with forceall=true, but I have left an issue to investigate further https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/39

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            if [ ! -s {params.irap_versions} ] ; then
+                echo "ERROR: Failed to retrieve {params.irap_versions} from singularity" >&2
+                exit 1
+            fi
```
If we use the rule as I suggested before, then this wouldn't be repeated here, and you would just use the output of that rule as input here.

pmb59 replied:
done

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            --geneNameFile "$geneNameFile" \
+            --source "${{i}}" \
+            | awk 'NR == 1; NR > 1 {{print $0 | "sort -n"}}' \
+            > $decoratedFile.swp
```
```suggestion
                --geneNameFile "$geneNameFile" \
                --source "${{i}}" \
                | awk 'NR == 1; NR > 1 {{print $0 | "sort -n"}}' \
                > $decoratedFile.swp
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        if [ $? -ne 0 ]; then
+            echo "ERROR: Failed to calculate normalized expressions for {wildcards.accession}" >&2
+            exit 1
+        fi
```
```suggestion
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        for p in {params.array_designs}
+        do  
+            echo "Array designs found in the xml config file: ${{p}}"
+            if [ -s {wildcards.accession}_"${{p}}-normalized-expressions.tsv.undecorated" ] ; then
+                echo "File {wildcards.accession}_${{p}}-normalized-expressions.tsv.undecorated exists for {wildcards.accession} and array_design ${{p}}"
+                cp {wildcards.accession}_"${{p}}-normalized-expressions.tsv.undecorated" {wildcards.accession}_"${{p}}-normalized-expressions.tsv.undecorated.unmerged"
+            else
+                echo "ERROR: File does not exist for {wildcards.accession} and array_design ${{p}}" >&2
+                exit 1
```
if `{wildcards.accession}_"${{p}}-normalized-expressions.tsv.undecorated"` would be an explicit output of the rule that generates it, it could be used as an input here (or a list of inputs), and reduce the amount of things here.

pmb59 replied:
This will be handled in issue https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/33

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        # check that qc dirs exist for each array design
+        for p in {params.array_designs}
+        do
+            if [ -d "./qc/{wildcards.accession}_${{p}}_QM" ]; then
+                echo " ./qc/{wildcards.accession}_${{p}}_QM found"
+            else
+                echo "ERROR: ./qc/{wildcards.accession}_${{p}}_QM not found"
+                exit 1
+            fi
```
again, I wonder if it is not better to declare these are outputs of the rule that generates them and consume them here are inputs. Then we don't need these checks.

pmb59 replied:
This will be handled in issue https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/33

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        if [ $? -ne 0 ]; then
+            echo "ERROR: Generate temp decorated files for {wildcards.accession}" >&2
+            exit 1
+        fi
```
```suggestion
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            if [ $? -ne 0 ]; then
+                echo "ERROR: Could not find array design: $arrayDesign" >&2
+                exit 1
+            fi
```
If we want to signal that the error is here, we could add an echo before or after the command signalling it started and/or finished.

pmb59 replied:
5f6f770

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+            if [ $? -ne 0 ]; then
+                echo "ERROR: Could not find array design: $arrayDesign" >&2
+                exit 1
+            fi
```
```suggestion
```

pmb59 replied:
5f6f770

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+	        echo "ERROR: Failed to calculate analytics for {wildcards.accession}" >&2
+	        rm -rf *.png *-analytics.tsv.undecorated
+	        exit 1
+        fi
```
```suggestion
```

pmb59 replied:
64cfd7e

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        echo $TMPDIR"/tmp"
+        if [ ! -d "$TMPDIR"/tmp ]; then
+            mkdir $TMPDIR/tmp
+        fi
```
lets try to reduce all of these to what I suggested, or at least replace them with a snakemake block that is used everywhere.

pmb59 replied:
Currently we depend on a custom temp_dir location in codon (see comment above), but I have moved basic logic to the common routines script f678934

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+                if [ $? -ne 0 ]; then
+                    echo "ERROR: no non-NA p-values found in {wildcards.accession}_${{p}}-analytics.tsv.undecorated" >&2
+                    exit 1
+                fi
```
```suggestion
```

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+                echo "ERROR: Failed to round to one decimal place log2fold changes in {params.analytics} " >&2                                                                                                 
+                rm -rf {params.intermediate_rounded}                                                                                                                                                           
+                exit 1                                                                                                                                                                                         
+        fi                                                                                                                                                                                                     
```
```suggestion                                                                                                                                                                                                  
```

pmb59 replied:
355c51c

pcm32 commented on Snakefile:
```diff
@@ -420,92 +637,1344 @@ rule atlas_experiment_summary:
+        if [ $? -ne 0 ]; then
+            echo "ERROR: Failed to decorate {wildcards.accession} and array {wildcards.array_design}" >&2
+            exit 1
+        fi
```
```suggestion
```

pcm32 commented on Snakefile-reprocess:
```diff
@@ -1,26 +1,175 @@
+        #        metric_link_coexp=True
+        #        outputs.extend(expand(f"{config['accession']}-{m}-coexpressions.tsv.gz"))
+        #if metric_link_coexp == True:
+        #    outputs.extend(expand(f"{config['accession']}-coexpressions.tsv.gz" ))
```
```suggestion
```

pcm32 replied:
if we don't want it I mean, things are fairly long already.

pcm32 commented on bin/arrayNormalization.R:
```diff
@@ -0,0 +1,391 @@
+		else if(mode == "lumi") {
+
+			illuminaArray( files, outFile )
+            
```
```suggestion
			illuminaArray( files, outFile )
```

pcm32 commented on bin/arrayNormalization.R:
```diff
@@ -0,0 +1,391 @@
+		if(mode == "agil1" | mode == "agil2") {
+	
+			agilentArray( files, mode, assayNames, outFile, miRBaseFile )
+
```
```suggestion
			agilentArray( files, mode, assayNames, outFile, miRBaseFile )
```

pcm32 commented on bin/arrayNormalization.R:
```diff
@@ -0,0 +1,391 @@
+
+			affymetrixArray( files, assayNames, outFile )
+
+		}
```
```suggestion
		else if(mode == "affy") {
			affymetrixArray( files, assayNames, outFile )
		}
```

pcm32 commented on bin/arrayNormalization.R:
```diff
@@ -0,0 +1,391 @@
+		
+		print("Reading Agilent 2-colour data")
+		dataSet <- try({read.maimages(files, source="agilent")})
+	}
```
```suggestion
	if(mode == "agil1") {
		print("Reading Agilent 1-colour data")
		# for 1-colour Agilent data we only have a green channel.
		dataSet <- try({read.maimages(files, source="agilent", green.only=TRUE)})
	}
	else if(mode == "agil2") {
		print("Reading Agilent 2-colour data")
		dataSet <- try({read.maimages(files, source="agilent")})
	}
```

pcm32 commented on bin/arrayNormalization.R:
```diff
@@ -0,0 +1,391 @@
+	if(mode == "agil1") {
+		
+		# Between-array normalization for 1-colour.
+		normData <- normalizeBetweenArrays(dataSet, method="quantile")
```
```suggestion
		# Between-array normalization for 1-colour.
		normData <- normalizeBetweenArrays(dataSet, method="quantile")
```

pcm32 commented on bin/arrayQC.R:
```diff
@@ -0,0 +1,312 @@
+	if(exptType == "affy") {
+   
+ 		dataSet <- affymetrixQC( annotationFile )
+	}
```
```suggestion
	if(exptType == "affy") {
 		dataSet <- affymetrixQC( annotationFile )
	}
```

pcm32 commented on bin/arrayQC.R:
```diff
@@ -0,0 +1,312 @@
+
+		dataSet <- agilentQC(annotationFile, exptType, miRBaseFile)
+
+	}
```
```suggestion
	else if(exptType == "lumi") {
		dataSet <- illuminaQC( annotationFile )
	}
	# Agilent data.
	# One-colour or Two color Agilent data
	else if(exptType == "agil1" | exptType == "agil2") {
		dataSet <- agilentQC(annotationFile, exptType, miRBaseFile)
	}
```

pcm32 commented on bin/arrayQC.R:
```diff
@@ -0,0 +1,312 @@
+	# 					reporttitle = reportTitle)
+	#-------------------------------------------------- 
+	
+    ##########################
+    # FIXME: Temporary workaround while arrayQualityMetrics package has a bug.
+    # Courtesy of Mike Smith on the Bioconductor support site: https://support.bioconductor.org/p/76626/#76640
+    preparedData <- prepdata(expressionset = dataSet, do.logtransform = TRUE, intgroup = factorColName)
+    bo = aqm.boxplot(preparedData)
+    de = aqm.density(preparedData)
+    hm = aqm.heatmap(preparedData)
+    pca = aqm.pca(preparedData)
+    ma = aqm.maplot(preparedData)
+    qm = list("Boxplot" = bo, "Density" = de, "Heatmap" = hm, "PCA" = pca, "MAplot" = ma)
+    arrayQCdata <- aqm.writereport( 
+        modules = qm, 
+        arrayTable = preparedData$pData, 
+        reporttitle = reportTitle, 
+        outdir = outDir
+    )
+    # End temporary workaround
+    ##########################
```
Is this ours (recent) or has been there for years probably?

pmb59 replied:
indeed, this chunk has been inherited from our predecessors

pcm32 commented on bin/arrayQC.R:
```diff
@@ -0,0 +1,312 @@
+
+	return(dataSet)
+
+}
```
```suggestion
	}
	return(dataSet)
}
```

pcm32 commented on bin/check_na_pvals.R:
```diff
@@ -0,0 +1,18 @@
+nonNAcounts <- sapply( pvalueColnames, function( x ) {
+    
+    length( which( !is.na( analyticsResults[[ x ]] ) ) )
+} )
```
```suggestion
args <- commandArgs( TRUE )
analyticsFilename <- args[ 1 ]
analyticsResults <- read.delim( analyticsFilename, header=TRUE )
pvalueColnames <- colnames( analyticsResults )[ grep( "p.value", colnames( analyticsResults ) ) ]
nonNAcounts <- sapply( pvalueColnames, function( x ) {
    length( which( !is.na( analyticsResults[[ x ]] ) ) )
} )
```

pcm32 commented on envs/differential-stats.yaml:
```diff
@@ -0,0 +1,14 @@
+channels:
```
Maybe, now that this is well established, we should aim to pin most versions to the current ones in the working environments. But we can maybe do that on a separate PR.

pcm32 replied:
Or at least avoid that they go beyond the next major version.

pmb59 replied:
See issue https://github.com/ebi-gene-expression-group/bulk-recalculations/issues/34

pcm32 commented on run_sorting_hat_test_data.sh:
```diff
@@ -21,6 +56,13 @@ CONDA_PREFIX_LINE="--conda-prefix $SN_CONDA_PREFIX"
 export LOG_PATH=${LOG_PATH:-$( pwd )/sorting.log}
 USUAL_SM_ERR_OUT=${USUAL_SM_ERR_OUT:-$( pwd )/snakemake.log}
 
+# isl db
+ORACLE_BASE=
+ORACLE_HOME=
+PYTHON_USER=
+PYTHON_CONNECT_STRING=
+PYTHON_PASSWORD=
```
If this is meant to be like a sample script and the one with all the real values is stored elsewhere, maybe we can then rename this to `*.sample.sh`.

pcm32 replied:
Otherwise I would suggest that this script can read these vars from outside and then complain if not set.

pmb59 replied:
thanks, I agree a859e53

pcm32 reviewed (commented):
A titanic effort @pmb59, well done. Looks very good in general, I only have some minor concerns for some patterns that might be adding more code than we need in the bash parts (ie. using files as params instead of as output inputs means that we need to be checking manually certain things that snakemake will check for you by default). But if those are major changes we could as well open issues and fix those iteratively later, specially once we have tests for this whole thing. There is a lot of lines that we can get rid of due to the `set -e` not allowing us to reach certain inherited checks, in some cases it might be useful to signal where the process ended though for the operator. We can discuss this on a call if it is easier.

pcm32 commented:
Maybe we could also update the README to reflect what can this repo do after these changes.

pmb59 committed:
db param->ISL db param

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
uncomment sys.exit

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
add f string

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
convert to f-strings

pmb59 committed:
convert to f-strings

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
convert to f-strings

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
convert to f-strings

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
convert to f-strings

pmb59 committed:
convert to f-strings

pmb59 committed:
remove from local rules those that involve heavy data transfer

pmb59 committed:
remove avoidable params

pmb59 committed:
delete unreachable code due to the `set -e`.

pmb59 committed:
delete unreachable code due to the `set -e`.

pmb59 committed:
rename sorting hat script

pmb59 committed:
update sample.sh sorting-hat script

pmb59 committed:
add indentation

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
pass avail custom memory (depends on workflow `attempt` number) to JVM for Ammonite (amm -s) REPL

pmb59 committed:
rm comment #test -s "

pmb59 committed:
Update Snakefile

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
replace param -> output

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
edit output claimed by rule create_tracks_symlinks

pmb59 committed:
rm block of code unreachable due to  set -e

pmb59 committed:
rm block of code unreachable due to set -e

pmb59 committed:
rm unreachable code block due to set -e

pmb59 committed:
Update Snakefile

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
rm unreachable block of code due to set -e

pmb59 committed:
rm unreach. block of code

pmb59 committed:
rm unused block of code

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
rm unused block of code

pmb59 committed:
remove unused code block

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
delete comments

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
add indents

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayNormalization.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayNormalization.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayNormalization.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayNormalization.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayNormalization.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayQC.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayQC.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/arrayQC.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update bin/check_na_pvals.R

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update README.md

pmb59 committed:
Update README.md

pmb59 committed:
add rule get_irap_versions_file

pmb59 committed:
reorder vars and cleanup in sorting hat script

pmb59 committed:
Update README.md

pmb59 committed:
set up a random temp dir name if not defined

pmb59 committed:
move basic logic to bin/reprocessing_routines.sh

pmb59 committed:
remove duplicated line
